lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -723,43 +738,79 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
-                    loop (fetchThem $ fetcher [identityUrlBuilder])
+                    queue <- STM.atomically $ newTBQueue queueSize
+                    concurrently_
+                        (enqueueMetadata queue)
+                        (dequeueMetadata manager [identityUrlBuilder] queue)
+
                 FetchSMASH (unSmashServer -> uri) -> do
                     STM.atomically $ writeTVar gcStatus NotStarted
                     let getDelistedPools =
                             fetchDelistedPools trFetch uri manager
                     tid <- forkFinally
                         (gcDelistedPools gcStatus tr db getDelistedPools)
                         onExit
+                    queue <- STM.atomically $ newTBQueue queueSize
                     flip finally (killThread tid) $
-                        loop (fetchThem $ fetcher [registryUrlBuilder uri])
-        | otherwise -> traceWith tr MsgSMASHUnreachable
+                        concurrently_
+                            (enqueueMetadata queue)
+                            (dequeueMetadata manager [registryUrlBuilder uri] queue)
 
+        | otherwise ->
+            traceWith tr MsgSMASHUnreachable
   where
     trFetch = contramap MsgFetchPoolMetadata tr
-    fetchThem fetchMetadata = do
+
+    queueSize :: Natural
+    queueSize = 100
+
+    -- Must be:
+    --     - > 1
+    --     - <= queueSize
+    -- Should be:
+    --     - <= queueSize / 2
+    numInFlight :: Int
+    numInFlight = 20
+
+    enqueueMetadata
+        :: TBQueue (PoolId, StakePoolMetadataUrl, StakePoolMetadataHash)
+        -> IO ()
+    enqueueMetadata queue = forever $ do
         refs <- atomically (unfetchedPoolMetadataRefs 100)
-        successes <- fmap catMaybes $ forM refs $ \(pid, url, hash) -> do
-            fetchMetadata pid url hash >>= \case
-                Nothing -> Nothing <$ do
+        STM.atomically $ mapM_ (writeTBQueue queue) refs
+        when (null refs) $ do
+            traceWith tr $ MsgFetchTakeBreak blockFrequency
+            threadDelay blockFrequency
+
+    dequeueMetadata
+        :: Manager
+        -> [UrlBuilder]
+        -> TBQueue (PoolId, StakePoolMetadataUrl, StakePoolMetadataHash)
+        -> IO ()
+    dequeueMetadata manager strategies queue = forever $ do
+        refs <- STM.atomically $ do
+            -- First read is blocking, and others aren't. This ensures a cheap
+            -- pause of the thread when there's nothing to fetch, but it always
+            -- try to fetch at most 'numInFlight' at once without blocking if
+            -- there are less to fetch.
+            -- This could be the case when the queue has been emptied and
+            -- registrations are now coming one-by-one.
+            h <- readTBQueue queue
+            q <- catMaybes <$> replicateM (numInFlight - 1) (tryReadTBQueue queue)
+            pure (h:q)
+        forConcurrently_ refs $ \(pid, url, hash) -> do
```

Comment:
This will still have uneven bursts of requests, won't it?
But if the fetch time is down to one minute, then it's not really a problem.

Reply:
I see what you mean, we could actually make the consumer send queries one-by-one (asynchronously) and actually use the queue's size to limit the number of requests in flight :thinking: ...

Reply:
Done :+1:

Reply:
Whoops. Actually no, it's not that simple because now `unfetchedPoolMetadataRefs` gets called multiple times in a row before it's correctly updated.

Reply:
Okay, now done properly. After a full run, there's no attempts with a retry_count greater than 3, which sounds right.